pkg/controller/controller.go

```diff
@@ -325,11 +345,55 @@ func (c *Controller) syncHandler(key string) error {
 	if err != nil {
 		return err
 	}
+	// // Finally, we update the status block of the Function resource to reflect the
+	// // current state of the world
+	err = c.updateFunctionStatus(function, deployment)
+	if err != nil {
+		return err
+	}
 
 	c.recorder.Event(function, corev1.EventTypeNormal, SuccessSynced, MessageResourceSynced)
 	return nil
 }
 
+func (c *Controller) updateFunctionStatus(fn *faasv1.Function, deployment *appsv1.Deployment) error {
+	// NEVER modify objects from the store. It's a read-only, local cache.
+	// You can use DeepCopy() to make a deep copy of original object and modify this copy
+	// Or create a copy manually for better performance
+	var status metav1.ConditionStatus = "True"
+	if deployment.Status.UnavailableReplicas > 0 {
+		status = "False"
+	}
+	var replace = false
+	for _, condition := range fn.Status.Conditions {
+		if condition.Type == "Ready" && condition.Status != status {
+			replace = true
+		}
+	}
+	if fn.Status.UnavailableReplicas != deployment.Status.UnavailableReplicas || fn.Status.ObservedGeneration != fn.ObjectMeta.Generation {
+		replace = true
+	}
+	var err error
+	if replace {
+		fnCopy := fn.DeepCopy()
+		fnCopy.Status.Conditions = []faasv1.FunctionCondition{
+			{
+				Status: status,
+				Type:   "Ready",
+			},
+		}
+		fnCopy.Status.UnavailableReplicas = deployment.Status.UnavailableReplicas
+		fnCopy.Status.ObservedGeneration = fn.ObjectMeta.Generation
+		// If the CustomResourceSubresources feature gate is not enabled,
+		// we must use Update instead of UpdateStatus to update the Status block of the Function resource.
+		// UpdateStatus will not allow changes to the Spec of the resource,
+		// which is ideal for ensuring nothing other than resource status has been updated.
+		_, err = c.faasclientset.OpenfaasV1().Functions(fn.Namespace).UpdateStatus(fnCopy)
```

Comment:
One improvement that will make this easier to test and maintain is to split this into two methods, one that calculates the new stats and one that does the update. That will allow us to write a simple unit test that verifies how  the status is calculated. Having the tests that show how specific cases change (or dont) will make it much easier to maintain